### DISCUSSION TITLE
Reloading or Refreshing the Root Object

### DISCUSSION THREAD

**Discussion by GUI-Botox** [nasa/openmct#7911] [Community Q&A]
Is there a way to refresh the Root Object such that if a new object is added to the Root Object, the object tree will refresh? 

For example, having a plugin that is installed after Open MCT is running in the window which adds to the Root Object composition doesn't trigger a refresh of the Object Tree on the left navigation pane. However, searching for a telemetry object associated with installed plugin after Open MCT runtime shows that the plugin successfully added to the Root and the telemetry object is available. This is the case with installing the YAMCS telemetry plugin after Open MCT is running, then searching for a telemetry object in the search bar or console logging the Root Object after running an install command.

```
installButton.addEventListener('click', () => {
  openmct.install(YAMCSPlugin(YAMCSConfig));
  console.log('objects: ', openmct.objects);
});
refreshButton.addEventListener('click', async () => {
  const rootObject = await openmct.objects.getRoot();
  const reload = openmct.actions.getAction('reload');
  reload.invoke([rootObject]);
  console.log(
    'objects: ',
    openmct.objects,
    rootObject
  );
});
```

**Comment by akhenry** [comment]
Hi @GUI-Botox,

Thanks for reaching out. This looks like a limitation of our API. Although [dynamic composition providers](https://github.com/nasa/openmct/blob/master/API.md#composition-providers) are indeed supported, they are not supported at the root level, which is a miss on our part.

The example code you provided for reloading the root object is a great attempt at forcing a reload. The reload action you used is actually aimed at [reloading views](https://github.com/nasa/openmct/blob/master/src/plugins/reloadAction/ReloadAction.js#L38), rather than domain object models. The equivalent method for refreshing domain object models is [`openmct.objects.refresh(...)`](https://github.com/nasa/openmct/blob/master/src/api/objects/ObjectAPI.js#L636). That said, it doesn't work with root objects.

So, what I'll do is create an issue to fix our root object provider, to support adding and removing root objects after startup, and improve some of our documentation around this generally.

In the mean-time, the following non-API code will force a refresh of the tree and pick up any new roots - 

```JavaScript
openmct.layout.$refs.mctTree.loadRoot(); 
```

**Reply by ikirk** [reply]
I believe I am facing a similar problem to @GUI-Botox, but I am actually trying to refresh a domain object rather than the root object itself. The object's view gives the user the option to select which dictionary to use for incoming telemetry, after which the object's composition contains the telemetry measurements for the selected dictionary, which is retrieved from an external server. The composition loads correctly and contains the correct values, but the object tree in the Browse list doesn't update until the object is collapsed and re-expanded. I've tried the `openmct.objects.refresh(...)` method that you mentioned, but that doesn't visibly change anything. the work-around that you mentioned (`openmct.layout.$refs.mctTree.loadRoot()`) works but collapses all objects in the tree, which is an annoying behavior for the user.

gif of the behavior (notice that "Test Secret Missile" has an extra telemetry measurement "Fuel," which doesn't appear underneath "Telemetry Source" until after it's collapsed and re-expanded):
![chrome_XSLph3GxO0](https://github.com/user-attachments/assets/49cbd872-bb90-4a59-a82d-7480a544a2b3)

The `openmct.objects.refresh(...)` method is called on the Telemetry Source domain object inside of an event listener that is listening to changes in the select box, which also handles other changes that _are_ visibly occurring, so the composition collection is correct at the point where refresh is being called. Any help would be greatly appreciated.

**Reply by inad9300** [reply]
In case it helps inform your decisions about a potential future public API, I am currently doing the following to both load a new root and open it:

```js
const { mctTree } = openmct.layout.$refs;
await mctTree.loadRoot();
await mctTree.syncTreeOpenItems();
```

Even if I did not want to open the root, without the second operation, the root's expanding arrow stays where it was before loading the new root, so sometimes you would have to click twice to actually open it.

**Comment by inad9300** [comment]
@akhenry Is there any equivalent way to refresh telemetry charts and dashboards? I would like to programmatically trigger the refetching of both historical and real-time telemetry data with new parameters (set externally).

**Reply by inad9300** [reply]
It seems to work fine through `openmct.time.setBounds(openmct.time.getBounds())`.

**Reply by akhenry** [reply]
Yes, the ClearData action should do what you want.
1. Install the plugin in your index.html or index.js -
``` JavaScript
openmct.install(
  openmct.plugins.ClearData(
    ['table', 'telemetry.plot.overlay', 'telemetry.plot.stacked', 'example.imagery'],
    { indicator: true }
  )
);
```
2. Programmatically invoke the action
``` JavaScript
openmct.actions.getAction('clear-data-action').invoke(this.openmct.router.path); 
```
